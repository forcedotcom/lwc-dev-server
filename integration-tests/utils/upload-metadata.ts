import archiver from 'archiver';
import jsforce from 'jsforce';
import fs from 'fs';
import path from 'path';

interface UploadOptions {
    /**
     * Absolute path to the package folder (e.g., force-app/main/default).
     */
    packagePath: string;
    /**
     * Existing authenticated jsforce connection.
     */
    connection: jsforce.Connection;
    /**
     * Include all files in the package directory.
     */
    all?: boolean;
    /**
     * Include the `classes` directory from the package directory.
     */
    apex?: boolean;
}

/**
 * Uploads metadata (apex classes, custom objects, etc...) to Salesforce.
 *
 * A package.xml file is required in the package directory. See
 * https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/file_based_zip_file.htm
 * for documentation and examples.
 *
 * @param options Configuration options.
 */
export async function upload(options: UploadOptions) {
    const { packagePath } = options;
    const packageXmlPath = path.join(packagePath, 'package.xml');
    const packageName = 'testpkg';

    console.log(`uploading metadata from path ${packagePath}`);

    if (!fs.existsSync(packageXmlPath)) {
        throw new Error('Missing package.xml file in package directory');
    }

    const archive = archiver('zip');
    if (options.all) {
        archive.directory(options.packagePath, packageName);
    } else {
        archive.file(packageXmlPath, {
            name: `${packageName}/package.xml`
        });
        if (options.apex) {
            archive.directory(
                path.join(packagePath, 'classes'),
                `${packageName}/classes`
            );
        }
    }
    archive.finalize();

    await options.connection.metadata
        .deploy(archive, {
            allowMissingFiles: false,
            autoUpdatePackage: false,
            ignoreWarnings: true,
            rollbackOnError: true
        })
        // @ts-ignore type def must be wrong?
        .complete({ details: true })
        .then((result: jsforce.DeployResult) => {
            console.log(JSON.stringify(result, null, 2));
            if (!result.success) {
                throw new Error('Upload did not succeed');
            }
            console.log('done uploading metadata.');
        })
        .catch((e: Error) => {
            console.error(`error uploading metadata: ${e.message}`);
        });
}
