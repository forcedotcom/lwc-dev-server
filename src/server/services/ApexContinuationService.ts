import { CompileService } from '@webruntime/api';

const APEX_CONTINUATION_REGEX = /^(@salesforce\/apexContinuation)(?:\/([\w-]+\.[\w-]+(?:\.[\w-]+)?))?$/;

function matchesApexContinuationScopedModule(id: string) {
    return id && id.match(APEX_CONTINUATION_REGEX);
}

/**
 * Rollup plugin to resolve @salesforce/apexContinuation
 *
 */
function plugin() {
    return {
        name: 'rollup-plugin-apex-continuation',

        resolveId(id: string) {
            return matchesApexContinuationScopedModule(id) ? id : null;
        },
        load(id: any) {
            const idParts = matchesApexContinuationScopedModule(id);
            if (idParts) {
                throw new Error(
                    'Preview of component using Apex Continuation is not supported'
                );
            }
            return null;
        }
    };
}

export class ApexContinuationService implements CompileService {
    getPlugin() {
        return plugin();
    }

    async initialize() {}

    shutdown() {}
}
