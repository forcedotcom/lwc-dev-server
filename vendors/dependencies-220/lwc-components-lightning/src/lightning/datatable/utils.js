export const isObjectLike = function(value) {
    return typeof value === 'object' && value !== null;
};

const proto = {
    add(className) {
        if (typeof className === 'string') {
            this[className] = true;
        } else {
            Object.assign(this, className);
        }
        return this;
    },
    invert() {
        Object.keys(this).forEach(key => {
            this[key] = !this[key];
        });
        return this;
    },
    toString() {
        return Object.keys(this)
            .filter(key => this[key])
            .join(' ');
    },
};

export const classSet = function(config) {
    if (typeof config === 'string') {
        const key = config;
        config = {};
        config[key] = true;
    }
    return Object.assign(Object.create(proto), config);
};

export const isPositiveInteger = function(value) {
    return /^\d+$/.test(value);
};

export const clamp = function(num, min, max) {
    return num <= min ? min : num >= max ? max : num;
};

export function normalizePositiveIntegerAttribute(attrName, value, fallback) {
    if (isPositiveInteger(value)) {
        return parseInt(value, 10);
    }
    // eslint-disable-next-line no-console
    console.warn(
        `The attribute "${attrName}" value passed in is incorrect.
            "${attrName}" value should be an integer >= 0.`
    );
    return fallback;
}
