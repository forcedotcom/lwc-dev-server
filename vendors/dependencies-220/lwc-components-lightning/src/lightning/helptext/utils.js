export function getBubbleAlignAndPosition(
    triggerBoundingClientRect,
    bubbleBoundingClientRect,
    defaultAlign,
    shiftAmounts,
    availableHeight,
    availableWidth,
    xOffset,
    yOffset
) {
    let align = {
        horizontal: defaultAlign.horizontal,
        vertical: defaultAlign.vertical,
    };
    let positionAt = { top: null, right: null, bottom: null, left: null };

    const bubbleOverflows = getBubbleOverflows(
        triggerBoundingClientRect,
        bubbleBoundingClientRect,
        shiftAmounts,
        availableWidth,
        availableHeight
    );

    // evaluate where the bubble should be positioned
    const alignAndPosition = calculateAlignAndPosition(
        align,
        positionAt,
        bubbleOverflows,
        triggerBoundingClientRect,
        bubbleBoundingClientRect,
        availableWidth,
        availableHeight
    );

    align = alignAndPosition.alignment;
    positionAt = alignAndPosition.positioning;

    const result = { align };

    // assign default values for position bottom & left based on trigger element if needed
    // - default anchor point of popover is bottom center attached to trigger element's top center
    positionAt.bottom =
        positionAt.top || positionAt.top === 0
            ? null
            : availableHeight - triggerBoundingClientRect.top;

    // set the left positioning default according to vertical alignment when needed
    let defaultLeft =
        align.vertical === 'center'
            ? triggerBoundingClientRect.right
            : triggerBoundingClientRect.left;

    // don't use the default if we already have a value
    if (positionAt.left) {
        defaultLeft = positionAt.left;
    }

    positionAt.left =
        positionAt.right || positionAt.right === 0 ? null : defaultLeft;

    const shiftByVertical =
        align.vertical === 'center' ? 0 : shiftAmounts.vertical;
    let shiftByHorizontal =
        align.horizontal === 'center' ? 0 : shiftAmounts.horizontal;

    // Change horizontal shift value to opposite value (negative or positive)
    // :: needed to push the bubble away from the trigger instead of into it when positioned on left or right
    if (align.vertical === 'center') {
        shiftByHorizontal *= -1;
    }

    // apply calculated position values
    result.top = positionAt.top
        ? positionAt.top + shiftByVertical + yOffset + 'px'
        : positionAt.top;
    result.right = positionAt.right
        ? positionAt.right - shiftByHorizontal - xOffset + 'px'
        : positionAt.right;
    result.bottom = positionAt.bottom
        ? positionAt.bottom + shiftByVertical - yOffset + 'px'
        : positionAt.bottom;
    result.left = positionAt.left
        ? positionAt.left - shiftByHorizontal + xOffset + 'px'
        : positionAt.left;

    return result;
}

export function getNubbinShiftAmount(nubbinComputedStyles, triggerWidth) {
    // calculate smallest positive value of horizontal nubbin distance, right or left
    // - the nubbin is the pointy element on the bubble
    const nubbinShiftLeft = parseInt(nubbinComputedStyles.left, 10) || -1;
    const nubbinShiftRight = parseInt(nubbinComputedStyles.right, 10) || -1;

    // check which measurement is the lesser of the two (closest to edge)
    let nubbinShift =
        nubbinShiftLeft < nubbinShiftRight ? nubbinShiftLeft : nubbinShiftRight;

    // use the positive, greater than zero, shift value
    if (nubbinShift < 0 && nubbinShiftLeft < 0 && nubbinShiftRight > 0) {
        nubbinShift = nubbinShiftRight;
    }
    if (nubbinShift < 0 && nubbinShiftRight < 0 && nubbinShiftLeft > 0) {
        nubbinShift = nubbinShiftLeft;
    }

    return {
        horizontal: nubbinShift - (triggerWidth / 2), // prettier-ignore
        vertical: parseInt(nubbinComputedStyles.height, 10),
    };
}

//
// Utility functions (for reduced complexity)
//
function getBubbleOverflows(
    triggerBoundingClientRect,
    bubbleBoundingClientRect,
    shiftAmounts,
    availableWidth,
    availableHeight
) {
    const bubbleOverflows = {};

    // evaluate in which directions the bubble overflows

    // is the bubble overflowing if positioned above the trigger?
    bubbleOverflows.top =
        triggerBoundingClientRect.top -
            (bubbleBoundingClientRect.height + shiftAmounts.vertical) <
        0;

    // is the bubble overflowing if halfway positioned above the trigger?
    // :: useful for vertical center calculation
    bubbleOverflows.topCenter =
        triggerBoundingClientRect.top - bubbleBoundingClientRect.height / 2 < 0;

    // is the bubble overflowing if positioned below the trigger?
    bubbleOverflows.bottom =
        triggerBoundingClientRect.bottom +
            bubbleBoundingClientRect.height +
            shiftAmounts.vertical >
        availableHeight;

    // is the bubble overflowing if positioned to the right of the trigger?
    bubbleOverflows.right =
        triggerBoundingClientRect.left + bubbleBoundingClientRect.width >
        availableWidth;

    // is the bubble overflowing if halfway positioned to the right of the trigger?
    // :: useful for horizontal center calculation
    bubbleOverflows.rightCenter =
        triggerBoundingClientRect.left + bubbleBoundingClientRect.width / 2 >
        availableWidth;

    // is the bubble overflowing if positioned to the left of the trigger?
    bubbleOverflows.left =
        triggerBoundingClientRect.right - bubbleBoundingClientRect.width < 0;

    // is the bubble overflowing if halfway positioned to the left of the trigger?
    // :: useful for horizontal center calculation
    bubbleOverflows.leftCenter =
        triggerBoundingClientRect.right - bubbleBoundingClientRect.width / 2 <
        0;

    return bubbleOverflows;
}

function calculateAlignAndPosition(
    align,
    positionAt,
    bubbleOverflows,
    triggerBoundingClientRect,
    bubbleBoundingClientRect,
    availableWidth,
    availableHeight
) {
    let bubbleIsVerticallyCentered = false;

    // if enough space to be vertically centered from top
    if (bubbleOverflows.top && !bubbleOverflows.topCenter) {
        align.vertical = 'center';

        // set the bubble to be vertically centered on the trigger
        // top position of the bubble to match the following formula:
        //  <bottom of trigger> - <half the height of trigger> - <half the height of the bubble>
        positionAt.top =
            triggerBoundingClientRect.bottom -
            triggerBoundingClientRect.height / 2 -
            bubbleBoundingClientRect.height / 2;

        bubbleIsVerticallyCentered = true;
        // if overflows upwards show below trigger
    } else if (bubbleOverflows.top) {
        align.vertical = 'top';
        positionAt.top = triggerBoundingClientRect.bottom;
    }

    // if overflows downward show above the trigger
    if (bubbleOverflows.bottom) {
        align.vertical = 'bottom';
        positionAt.bottom = availableHeight - triggerBoundingClientRect.top;
    }

    // if vertically centered and overflows left then show on right
    if (bubbleIsVerticallyCentered && bubbleOverflows.left) {
        align.horizontal = 'left';
        positionAt.left = triggerBoundingClientRect.right;
        // if overflows to the left show on right
    } else if (bubbleOverflows.left) {
        align.horizontal = 'left';
        positionAt.left = triggerBoundingClientRect.left;
    }

    // if vertically centered and overflows right then show on left
    if (bubbleIsVerticallyCentered && bubbleOverflows.right) {
        align.horizontal = 'right';
        positionAt.right = availableWidth - triggerBoundingClientRect.left;
        // if overflows to the right show on left
    } else if (bubbleOverflows.right) {
        align.horizontal = 'right';
        positionAt.right = availableWidth - triggerBoundingClientRect.right;
    }

    // only horizontally center bubble if it would overflow to the right or left
    if (
        bubbleOverflows.left &&
        bubbleOverflows.right &&
        !bubbleOverflows.leftCenter &&
        !bubbleOverflows.rightCenter
    ) {
        align.horizontal = 'center';

        // set the bubble to be horizontally centered on the trigger
        // left position of the bubble to match the following formula:
        //  <left edge of trigger> - <half the width of trigger> - <half the width of the bubble>
        positionAt.left =
            triggerBoundingClientRect.left +
            triggerBoundingClientRect.width / 2 -
            bubbleBoundingClientRect.width / 2;

        positionAt.right = null;
    }

    return {
        alignment: align,
        positioning: positionAt,
    };
}
