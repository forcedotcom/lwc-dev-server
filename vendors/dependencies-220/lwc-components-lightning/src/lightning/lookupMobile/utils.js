/**
 * Given a record, field, and objectInfos, returns a map of field attributes.
 *
 * Returned schema -
 * {
 *      inlineHelpText: {String},
 *      isRequired: {Boolean},
 * }
 *
 * @param {Object} record - A record representation.
 * @param {String} fieldName - A field api name.
 * @param {Object} objectInfos - A map of objectInfos.
 * @return {Object} A map of field attributes.
 */
function getFieldProperties(record, fieldName, objectInfos) {
    if (
        Object.keys(record || {}).length === 0 ||
        !fieldName ||
        Object.keys(objectInfos || {}).length === 0
    ) {
        return {
            fieldLevelHelp: null,
            isRequired: false,
        };
    }

    const objectApiName = record.apiName;
    const objectInfo = objectInfos[objectApiName];
    if (!objectInfo) {
        throw new Error(`ObjectInfo [${objectApiName}] was not found`);
    }
    const fieldInfo = objectInfo.fields[fieldName];
    if (!fieldInfo) {
        throw new Error(`Field [${fieldName}] was not found`);
    }
    return {
        fieldLevelHelp: fieldInfo.inlineHelpText,
        isRequired: fieldInfo.required,
    };
}

export { getFieldProperties };
