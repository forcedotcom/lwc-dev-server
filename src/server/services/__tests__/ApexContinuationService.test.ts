import { ApexContinuationService } from '../ApexContinuationService';
import { getLatestVersion } from '@webruntime/server/dist/commonjs/utils/utils';
import { PublicConfig } from '@webruntime/api';

jest.mock('@webruntime/server/dist/commonjs/utils/utils');

describe('ApexContinuationService', () => {
    let config: PublicConfig;

    beforeEach(() => {
        (getLatestVersion as jest.Mock).mockReturnValue('158104e2eb');
    });

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('returns a rollup plugin', () => {
        const service = new ApexContinuationService();
        expect(service.getPlugin()).toBeTruthy();
    });

    describe('rollup plugin', () => {
        describe('resolveId', () => {
            it('should resolve paths with @salesforce/apexContinuation', () => {
                const service = new ApexContinuationService();
                const plugin = service.getPlugin();

                const resolved = plugin.resolveId(
                    '@salesforce/apexContinuation/SampleContinuationClass.startRequest'
                );
                expect(resolved).toBe(
                    '@salesforce/apexContinuation/SampleContinuationClass.startRequest'
                );
            });

            it('should resolve to null for paths not starting with @salesforce/apexContinuation', () => {
                const service = new ApexContinuationService();
                const plugin = service.getPlugin();

                const resolved = plugin.resolveId(
                    '@salesforce/apex/SampleContinuationClass.startRequest'
                );
                expect(resolved).toBeNull();
            });
        });

        describe('load', () => {
            it('should throw error for @salesforce/apexContinuation imports', () => {
                (getLatestVersion as jest.Mock).mockReturnValue('158104e2eb');

                const service = new ApexContinuationService();
                const plugin = service.getPlugin();
                function loadApexContinuation() {
                    plugin.load(
                        '@salesforce/apexContinuation/SampleContinuationClass.startRequest'
                    );
                }
                expect(loadApexContinuation).toThrowErrorMatchingSnapshot();
            });

            it('should return null instead of throwing error for other than @salesforce/apexContinuation imports', () => {
                const service = new ApexContinuationService();
                const plugin = service.getPlugin();
                const resolved = plugin.load(
                    '@salesforce/apex/SampleContinuationClass.startRequest'
                );

                expect(resolved).toBeNull();
            });
        });
    });
});
