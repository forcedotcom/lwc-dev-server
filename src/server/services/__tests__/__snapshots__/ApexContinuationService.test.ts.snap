// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApexContinuationService rollup plugin load should throw error for @salesforce/apexContinuation imports 1`] = `"You can't preview a component using Apex Continuation. Test the component directly in the org"`;
